config/src/main/java/com/typesafe/config/ConfigDocument.java

```diff
@@ -0,0 +1,40 @@
+package com.typesafe.config;
+
+/**
+ * An object parsed from the original input text, which can be used to
+ * replace individual values and exactly render the original text of the
+ * input.
+ *
+ * <p>
+ * Because this object is immutable, it is safe to use from multiple threads and
+ * there's no need for "defensive copies."
+ *
+ * <p>
+ * <em>Do not implement interface {@code ConfigNode}</em>; it should only be
+ * implemented by the config library. Arbitrary implementations will not work
+ * because the library internals assume a specific concrete implementation.
+ * Also, this interface is likely to grow new methods over time, so third-party
+ * implementations will break.
+ */
+public interface ConfigDocument {
+    /**
+     * Returns a new ConfigDocument that is a copy of the current ConfigDocument,
+     * but with the desired value set at the desired path. If the path exists, it will
+     * remove all duplicates before the final occurrence of the path, and replace the value
+     * at the final occurrence of the path. If the path does not exist, it will be added.
+     *
+     * @param path the path at which to set the desired value
+     * @param newValue the value to set at the desired path, represented as a string. This
+     *                 string will be parsed into a ConfigNode, and the text will be inserted
+     *                 as-is into the document, with leading and trailing whitespace removed.
```

Comment:
Hmm, I'm not entirely sure how parsing the whole concatenation would work if it's JSON. Since the rendered output of the node created by the single value parsing needs to be valid JSON, we'd need to modify the rendered result so it's valid JSON. If we're combining simple values that's not a huge deal (can just put quotes around the rendered output) but when we get to objects and arrays we run into a problem wherein we need to merge not just the values but the text contained within them to form a single object, which seems pretty complicated.

Sounds like throwing an error if there's surrounding whitespace might be a good way to go. I could also add a new Node type or make ConfigNodeSimpleValue multi-token, though, either of those two approaches shouldn't be difficult.